sqlmesh/utils/file.py

```diff
@@ -0,0 +1,129 @@
+        return f"{self.path}.lock"
+
+    def read(self) -> t.Optional[bytes]:
+        """Reads data from the file."""
+        if self._fs.exists(self.path):
+            if not self._is_locked:
+                ... # TODO: do whatever we think is right here
+            content = self._fs.cat(self.path)
+            self._original_contents = content[:]
+            return content
+        return None
+    
+    def write(self, content: bytes) -> None:
+        """Writes data within the transaction."""
+        if not self._is_locked:
+            # TODO: do whatever we think is right here
+            raise RuntimeError("Cannot write to a file without a lock.")
+        self._fs.pipe(self.path, content)
+
+    def rollback(self) -> None:
+        """Rolls back the transaction."""
+        if self._original_contents is not None:
+            self._fs.pipe(self.path, self._original_contents)
+
+    def acquire_lock(
+        self, blocking: bool = True, timeout: float = -1
+    ) -> bool:
+        """Acquires a lock on a key.
+        
+        Acquire a lock, blocking or non-blocking.
+
+        When invoked with the blocking argument set to True (the default), block until
+        the lock is unlocked, then set it to locked and return True.
+
+        When invoked with the blocking argument set to False, do not block. If a call
+        with blocking set to True would block, return False immediately; otherwise, set
+        the lock to locked and return True.
+
+        When invoked with the floating-point timeout argument set to a positive value,
+        block for at most the number of seconds specified by timeout and as long as the
+        lock cannot be acquired. A timeout argument of -1 specifies an unbounded wait.
+        If blocking is False, timeout is ignored. The stdlib would raise a ValueError.
+
+        The return value is True if the lock is acquired successfully, False if
+        not (for example if the timeout expired).
+        """
+        if self._is_locked:
+            # Make this idempotent for a single instance.
+            return True
+        lock_id = uuid4()
+        wait_time = 0
+        while self._fs.exists(self.lock_path):
+            if not blocking:
+                return False
+            time.sleep(1)
+            wait_time += 1
+            if timeout > 0 and wait_time > timeout:
+                return False
+        self._fs.pipe(self.lock_path, json.dumps(
+            # Add more fields here if needed.
+            {"gid": str(lock_id)},
+        ))
+        self._jitter()
+        lock_content = json.loads(self._fs.cat(self.lock_path))
+        acquired = lock_content["gid"] == str(lock_id)
+        if not acquired:
+            # This indicates contention from a concurrent process.
+            # Retrying will now result in a wait.
+            return self.acquire_lock(blocking, timeout)
+        self._is_locked = True
+        return True
+
+    def release_lock(self) -> None:
```

Comment:
you should only release a lock if you have a lock right? this could falsely kill someone else's lock